sktime/annotation/stray.py

```diff
@@ -0,0 +1,283 @@
+        start = int(max(np.floor(n * (1 - self.p)), 1))
+
+        ghat = [
+            0.0 if i < start else sum((J / (n4 - 1)) * gaps[i - J + 1])
+            for i in range(n)
+        ]
+
+        log_alpha = np.log(1 / self.alpha)
+        bound = np.Inf
+
+        for i in range(start, n):
+            if gaps[i] > log_alpha * ghat[i]:
+                bound = outlier_score[order][i - 1]
+                break
+
+        return np.where(outlier_score > bound)[0]
+
+    def _use_KNN(self, X, n):
+        """Find outliers using kNN distance with maximum gap.
+
+        Parameters
+        ----------
+        X : np.ArrayLike
+            Data for anomaly detection (time series).
+        n : int
+            The number of rows remaining in X when NA's are removed.
+
+        Returns
+        -------
+        dict of index of outliers and the outlier scores
+        """
+        if len(X.shape) == 1:
+            nbrs = NearestNeighbors(n_neighbors=self.k + 1).fit(X.reshape(-1, 1))
+            distances, _ = nbrs.kneighbors(X.reshape(-1, 1))
+        else:
+            nbrs = NearestNeighbors(
+                n_neighbors=n if self.k >= n else self.k + 1,
+                algorithm=self.knn_algorithm,
+            ).fit(X)
+            distances, _ = nbrs.kneighbors(X)
+
+        if self.k == 1:
+            d = distances[:, 1]
+        else:
+            diff = np.apply_along_axis(np.diff, 1, distances)
+            d = distances[range(n), np.apply_along_axis(np.argmax, 1, diff) + 1]
+
+        out_index = self._find_threshold(d, n)
+        return {"idx_outliers": out_index, "out_scores": d}
+
+    def _find_HDoutliers(self, X):
+        """Detect Anomalies in High Dimensional Data.
+
+        Parameters
+        ----------
+        X : np.ArrayLike
+            Data for anomaly detection (time series).
+
+        Returns
+        -------
+        dict of anomalies and their corresponding scores
+        """
+        r = np.shape(X)[0]
+        idx_dropna = np.array([i for i in range(r) if not np.isnan(X[i]).any()])  # tag
+        X_dropna = X[
+            idx_dropna,
+        ]  # FIXME: might move this somewhere else?
+
+        X_dropna = np.apply_along_axis(self.normalize, 0, X_dropna)
+
+        n = np.shape(X_dropna)[0]
+        outliers = self._use_KNN(X_dropna, n)
+
+        # adjusted back to length r, for missing data
+        slice_ = [True if i in outliers["idx_outliers"] else False for i in range(n)]
+        idx_outliers = idx_dropna[slice_]  # index values from 1:r
+        outlier_bool = np.array([1 if i in idx_outliers else 0 for i in range(r)])
+
+        list_scores = outliers["out_scores"].tolist()
+        outlier_scores = np.array(
+            [list_scores.pop(0) if i in idx_dropna else np.nan for i in range(r)]
+        )
+
+        return {
+            "outlier_scores": outlier_scores,
+            "outlier_bool": outlier_bool,
+        }
+
+    def _fit(self, X, y=None):
```

Comment:
```suggestion
    def _fit(self, X: npt.ArrayLike, y: npt.ArrayLike = None):
```

Reply:
should we really use `npt.ArrayLike`? In docstring we always write `pandas` so far :thinking: